### DISCUSSION TITLE
MongoDB nested object query

### DISCUSSION THREAD

**Discussion by DieserMerlin** [mikro-orm/mikro-orm#1219] [Q&A] [answered]
Hey,
I noticed that mikro-orm creates invalid mongodb queries for nested objects.

I have this entity (minified):
```typescript
@Entity({customRepository: () => UserRepository})
export class User {
  [EntityRepositoryType]?: UserRepository;

  @Property()
  twitchData!: TwitchData;
}
```

where TwitchData is an object holding information like username, twitchId etc...

When I use my custom repository to query the database against a specific twitchId, I use this approach:
```typescript
export class UserRepository extends EntitiyRepository<User> {
  getByTwitchId(twitchId: string): Promise<User|null> {
    return this.findOne({twitchData: {twitchId}});
  }
}
```

This generates the following query:

> db.getCollection('user').find({ **twitchData: { twitchId: '...' }** }, { session: undefined }).limit(1).toArray();

The problem is, since you can't query subdocuments in MongoDB collections with subobjects in the query, but with dot notation, this just returns null. In order for the query to actually find documents, it would need to use MongoDB's dot notation as follows:

> db.getCollection('user').find({ **'twitchData.twitchId': '...'** }, { session: undefined }).limit(1).toArray();

Did I miss something in the documentation?

**Comment by B4nan** [comment, marked as answer]
Started working on #1017, with that the go to solution will be using embeddables in object mode. 

Currently you need to query via dot notation. Due to the strict typings you need to use `expr` helper:

```ts
return this.findOne({ [expr('twitchData.twitchId')]: twitchId });
```

If you need just one level of nesting, you can use embeddables right ahead, and it should work fine - the issue I mentioned is about being able to use embeddables inside embeddables. 

```ts
@Embedddable()
export class TwitchData {
  // here you define the structure via `@Property` decorator as usual

  @Property()
  twitchId: number;

}

@Entity({customRepository: () => UserRepository})
export class User {
  [EntityRepositoryType]?: UserRepository;

  @Embedded({ entity: () => TwitchData, object: true })
  twitchData!: TwitchData;
}
```

https://mikro-orm.io/docs/embeddables/
https://mikro-orm.io/docs/embeddables/#storing-embeddables-as-objects

**Comment by DieserMerlin** [comment]
Thank you!
